compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -381,11 +381,13 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
     }
 
     fn suggest_self_or_self_ref(&mut self, err: &mut Diagnostic, path: &[Segment], span: Span) {
-        let is_assoc_fn = self.self_type_is_available();
+        if !self.self_type_is_available() {
```

Comment:
This probably causes a regression in the suggestions for the code:
```rust
struct Foo {
    i: i32,
}

impl Foo {
    fn needs_self() {
        this.i
    }
}
```

Which currently looks like:
```
error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find value `this` in this scope
 --> src/lib.rs:7:9
  |
7 |         this.i
  |         ^^^^ not found in this scope
  |
help: you might have meant to use `self` here instead
  |
7 |         self.i
  |         ~~~~
help: if you meant to use `self`, you are also missing a `self` receiver argument
  |
6 |     fn needs_self(&self) {
  |                   +++++
```

Can you check that it still works? Also useful to commit a test for it.

Reply:
This testcase still works, I didn't change the meaning of `suggest_self_or_self_ref`, I moved the check from L388 to beginning of this function, removed a variable `is_assoc_fn`.
 
I will add this code into tests.